hphp/util/asm-x64-intelxed.h

```diff
@@ -0,0 +1,1066 @@
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr, xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, effOperandSizeBits, dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst1(i, s_xedState, instr,
+                                    effOperandSizeBits, op);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr, const xed_encoder_operand_t& op,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op, effOperandSizeBits, dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst2(i, s_xedState, instr,
+                                    effOperandSizeBits, op_1, op_2);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr,
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, effOperandSizeBits,
+                                   dest()));
+  }
+
+  ALWAYS_INLINE
+  uint32_t xedEmit(xed_iclass_enum_t instr,
+                   const xed_encoder_operand_t& op_1,
+                   const xed_encoder_operand_t& op_2,
+                   const xed_encoder_operand_t& op_3,
+                   xed_uint_t effOperandSizeBits,
+                   CodeAddress destination) {
+    return xedEmitImpl(instr, destination,
+                        [&](xed_encoder_instruction_t* i) {
+                          xed_inst3(i, s_xedState, instr,
+                                    effOperandSizeBits, op_1, op_2, op_3);
+                        });
+  }
+
+  ALWAYS_INLINE
+  void xedEmit(xed_iclass_enum_t instr,
+               const xed_encoder_operand_t& op_1,
+               const xed_encoder_operand_t& op_2,
+               const xed_encoder_operand_t& op_3,
+               xed_uint_t effOperandSizeBits = 0) {
+    codeBlock.moveFrontier(xedEmit(instr, op_1, op_2, op_3, effOperandSizeBits,
+                                   dest()));
+  }
+
+public:
+#define INT3_SIZE   (sz::byte)
+#define UD2_SIZE    (sz::word)
+
+  void emitInt3s(int n) {
+    static uint8_t instr = 0;
+    if (n == 0) return;
+    if (instr == 0) {
+      xedEmit(XED_ICLASS_INT3, sz::byte, &instr);
```

Comment:
This isn't thread-safe. Something like this should do it:
```c++
static auto const instr = []{
  uint8_t tmp;
  xedEmit(XED_ICLASS_INT3, sz::byte, &tmp);
  return tmp;
}();
```

Reply:
Addressed in https://github.com/facebook/hhvm/pull/8164/commits/41f6fc4dee0b2a2b3e3ea52d1c583202b583425f